mono/utils/mono-threads.c

```diff
@@ -1775,6 +1775,39 @@ mono_threads_close_thread_handle (MonoThreadHandle *thread_handle)
 	mono_refcount_dec (thread_handle);
 }
 
+/*
+ * mono_threads_open_native_thread_handle:
+ *
+ *  Duplicate the handle. The handle needs to be closed by calling
+ *  mono_threads_close_native_thread_handle () when it is no longer needed.
+ */
+MonoNativeThreadHandle
+mono_threads_open_native_thread_handle (MonoNativeThreadHandle thread_handle)
+{
+#ifdef HOST_WIN32
+	BOOL success = FALSE;
+	HANDLE new_thread_handle = NULL;
+
+	g_assert (thread_handle && thread_handle != INVALID_HANDLE_VALUE);
+	success = DuplicateHandle (GetCurrentProcess (), thread_handle, GetCurrentProcess (), &new_thread_handle, 0, FALSE, DUPLICATE_SAME_ACCESS);
+	g_assertf (success, "Failed to duplicate thread handle");
+
+	return new_thread_handle;
+#else
+	return NULL;
+#endif
+}
+
+void
+mono_threads_close_native_thread_handle (MonoNativeThreadHandle thread_handle)
+{
+#ifdef HOST_WIN32
+	g_assert (thread_handle != INVALID_HANDLE_VALUE);
+	if (thread_handle)
```

Comment:
There is a sort of confusing sublte inconsistency between `NULL` and `INVALID_HANDLE_VALUE`.
I get it though -- `NULL` is the "real" invalid value, that could route through mono and back to here. `INVALID_HANDLE_VALUE` should never make it out of Win32-specific code.

You might assert that `DuplicateHandle` never returns `INVALID_HANDLE_VALUE`, at least on the success path, but that would be pointless really.

You might put in a comment acknowledging or explaining the different handling of `NULL` and `INVALID_HANDLE_VALUE`.

Reply:
Maybe just remove the assert about `INVALID_HANDLE_VALUE`.
Don't mention `INVALID_HANDLE_VALUE` at all.

Reply:
NULL is what's used by mono to detect invalid value or lack of value, so that's the most important one. Added check for INVALID_HANDLE_VALUE since it should never really happen, but is a valid error state mostly indicating an implementation error elsewhere in the code (using a different way to setup handle), so keeping it in the Win32 specific code adds value.